cmd/snap-update-ns/change.go

```diff
@@ -438,94 +453,42 @@ func neededChangesImpl(currentProfile, desiredProfile *osutil.MountProfile) []*C
-
-		// Reuse synthetic entries if their needed-by entry is desired.
-		// Synthetic entries cannot exist on their own and always couple to a
-		// non-synthetic entry.
-
-		// NOTE: Synthetic changes have a special purpose.
-		//
-		// They are a "shadow" of mount events that occurred to allow one of
-		// the desired mount entries to be possible. The changes have only one
-		// goal: tell snap-update-ns how those mount events can be undone in
-		// case they are no longer needed. The actual changes may have been
-		// different and may have involved steps not represented as synthetic
-		// mount entires as long as those synthetic entries can be undone to
-		// reverse the effect. In reality each non-tmpfs synthetic entry was
-		// constructed using a temporary bind mount that contained the original
-		// mount entries of a directory that was hidden with a tmpfs, but this
-		// fact was lost.
-		if current[i].XSnapdSynthetic() && desiredIDs[current[i].XSnapdNeededBy()] {
-			logger.Debugf("reusing synthetic entry %q", current[i])
-			reuse[dir] = true
-			continue
-		}
-
-		// Reuse entries that are desired and identical in the current profile.
-		if entry, ok := desiredMap[dir]; ok && current[i].Equal(entry) {
-			logger.Debugf("reusing unchanged entry %q", current[i])
-			reuse[dir] = true
-			continue
-		}
-
-		skipDir = strings.TrimSuffix(dir, "/") + "/"
-	}
-
-	logger.Debugf("desiredIDs: %v", desiredIDs)
-	logger.Debugf("reuse: %v", reuse)
-
 	// We are now ready to compute the necessary mount changes.
 	var changes []*Change
 
-	// Unmount entries not reused in reverse to handle children before their parent.
+	// Unmount entries not reused in reverse order, so that the most nested
+	// element is always processed first.
 	for i := len(current) - 1; i >= 0; i-- {
-		if reuse[current[i].Dir] {
-			changes = append(changes, &Change{Action: Keep, Entry: current[i]})
-		} else {
-			var entry osutil.MountEntry = current[i]
-			entry.Options = append([]string(nil), entry.Options...)
-			// If the mount entry can potentially host nested mount points then detach
-			// rather than unmount, since detach will always succeed.
-			shouldDetach := entry.Type == "tmpfs" || entry.OptBool("bind") || entry.OptBool("rbind")
-			if shouldDetach && !entry.XSnapdDetach() {
+		var entry osutil.MountEntry = current[i]
+		entry.Options = append([]string(nil), entry.Options...)
+		switch {
+		case entry.XSnapdSynthetic() && entry.Type == "tmpfs":
+			// Synthetic changes are rooted under a tmpfs, detach that tmpfs to
+			// remove them all.
+			if !entry.XSnapdDetach() {
+				entry.Options = append(entry.Options, osutil.XSnapdDetach())
+			}
+		case entry.XSnapdSynthetic():
+			// Consume all other syn ethic entries without emitting either a
+			// mount, unmount or keep change.  This relies on the fact that all
+			// synthetic mounts are created by a mimic underneath a tmpfs that
+			// is detached, as coded above.
+			continue
```

Comment:
we discussed whether this was covered, it seems to be, not sure how extensively but unit tests do hit this line